### PR TITLE
DHCP: Fix sorting of leases

### PR DESCRIPTION
**Pull request by vnxme**
**Issues:**
1. DHCPv6 leases are shown in a wrong order (e.g. ::00d0 above ::0010)
1. DHCPv6 leases cannot be sorted by another column (e.g. hostname)
1. DHCPv6 leases have an empty interface field (all lines)
1. DHCPv6 leases have an empty interface field when IPv6 address belongs to /120 subnet

**Reasons:**
1. DHCPv4/v6 leases pages use strnatcasecmp() and strcmp() for sorting by IP
1. No UI correspoding to this feature
1. $staticent['ipaddrv6'] should be used instead of $staticent['ipaddr'] when searching a static mapping
1. Incomplete preg pattern used to find possible IPv6 subnets

**Solutions:**
1. Use the updated ipcmp() function for sorting by IPv4/v6
1. Add sorting UI to DHCPv6 leases page
1. Fix a typo
1. Apply a consistent approach of using is_subnet{v4,v6} to find possible subnets

**Version:**
OPNsense  20.7.6

### CODE REVIEW

File: src/www/status_dhcp_leases.php [hunk 1]
```diff
@@ -246,15 +246,19 @@ function remove_duplicate($array, $field)
 
     $order = ( $_GET['order'] ) ? $_GET['order'] : 'ip';
 
-    usort($leases,
-        function ($a, $b) use ($order) {
+    if ($order === 'ip') {
```
Comment (fichtner) [inline src/www/status_dhcp_leases.php:249]: this looks a bit convoluted. If we pass $order top the usort function we only need one which does inline if/else on the function to use?

File: src/www/status_dhcpv6_leases.php [hunk 2]
```diff
@@ -307,9 +307,19 @@ function parse_duid($duid_string)
         }
     }
 
-    if ($_GET['order']) {
+    $order = ( $_GET['order'] ) ? $_GET['order'] : 'ip';
```
Comment (fichtner) [inline src/www/status_dhcpv6_leases.php:310]: this looks fishy -- some pages have weak handling of input arguments and their defaults

Comment (fichtner) [reply src/www/status_dhcpv6_leases.php:310]: hmm, ok it emulates the dhcp v4 leases page. now we have two problems :D

File: src/etc/inc/interfaces.inc [hunk 3]
```diff
@@ -3772,21 +3772,20 @@ function guess_interface_from_ip($ipaddress)
         /* create a route table we can search */
         exec("/usr/bin/netstat -rnWf inet", $output, $ret);
         foreach ($output as $line) {
-            if (preg_match("/^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+\/[0-9]+[ ]+link[#]/", $line)) {
```
Comment (fichtner) [inline src/etc/inc/interfaces.inc:3775]: interfaces.inc changes should probably not be part of this PR as it is used outside of dhcp leases page as well. oddly enough it is not used in the v4 leases page.

File: src/www/status_dhcp_leases.php (hunk 1 above)
Comment (vnxme) [reply src/www/status_dhcp_leases.php:249]: Not sure whether I understand your idea right. In commit 2adf500 I put the $order condition inside the usort function and it looked quite nice. Then I realised that we can eliminate this condition being checked multiple times while sorting by means of splitting the single usort function into two ones: a simple one for sorting by IP (no need to even have $order inside the function namespace) and an extra one for sorting by arbitrary field chosen by the user. The latter approach is what I implemented in commit a012f89.

File: src/www/status_dhcpv6_leases.php (hunk 2 above)
Comment (vnxme) [reply src/www/status_dhcpv6_leases.php]: Yes, I copied it from the DHCPv4 leases page. Suggested a more secure way for both pages in commit 43427bb.

File: src/etc/inc/interfaces.inc (hunk 3 above)
Comment (vnxme) [reply src/etc/inc/interfaces.inc:3775]: I have nothing against pulling these changes separately. However, some DHCPv6 leases (e.g. /120 ones) will not be shown correctly unless the interfaces.inc file is fixed.

File: src/www/status_dhcpv6_leases.php [hunk 4]
```diff
@@ -430,22 +453,21 @@ function parse_duid($duid_string)
...
+                    <th class="act_sort" data-field="start"><?=gettext("Start"); ?></th>
+                    <th class="act_sort" data-field="end"><?=gettext("End"); ?></th>
+                    <th class="act_sort" data-field="online"><?=gettext("Online"); ?></th>
+                    <th class="act_sort" data-field="act"><?=gettext("Lease Type"); ?></th>
                     <th class="text-nowrap"></th>
                 </tr>
               </thead>
               <tbody>
 <?php
-              $mac_man = json_decode(configd_run("interface list macdb json"), true);
```
Comment (fichtner) [inline src/www/status_dhcpv6_leases.php:448]: We should probably show the info below the mac address instead of deleting the unused variable.

File: src/www/status_dhcpv6_leases.php [hunk 5]
```diff
@@ -454,7 +476,7 @@ function parse_duid($duid_string)
                     foreach ($config['dhcpdv6'] as $dhcpif => $dhcpifconf) {
                         if (isset($dhcpifconf['staticmap'])) {
                             foreach ($dhcpifconf['staticmap'] as $staticent) {
-                                if ($data['ip'] == $staticent['ipaddr']) {
+                                if ($data['ip'] == $staticent['ipaddrv6']) {
```
Comment (fichtner) [inline src/www/status_dhcpv6_leases.php:479]: looks like a fix... can we separate the $mac_man use and this into a separate PR?

File: src/etc/inc/interfaces.inc (hunk 3 above)
Comment (fichtner) [reply src/etc/inc/interfaces.inc:3775]: I'm saying its easier to pull in fixes first before working on GUI representation...

File: src/www/status_dhcp_leases.php (hunk 1 above)
Comment (fichtner) [reply src/www/status_dhcp_leases.php:257]: This inflates size about a factor of four compared to when the code was when this lease cleanup started. I understand that adding code is necessary to improve functionality but as a leaf page we don't need to prematurely optimise it.

**Comment by vnxme** [comment]
OK, I will prepare a few separate PRs and list them below.
